lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs

```diff
@@ -128,10 +150,189 @@ hashSeed :: ScrubbedBytes -> ScrubbedBytes
+                    pure Nothing
+                CryptoPassed plaintext ->
+                    Just <$> deserialiseDerPath plaintext
+        _ -> fail $ "decodeDerivationPath: Incorrect map length: " ++ show len
+  where
+    passphrase = hdPassphrase rootKey
+    deserialiseDerPath :: ByteString -> CBOR.Decoder s (Index 'Hardened 'AccountK, Index 'Soft 'AddressK)
+    deserialiseDerPath bytes =
+        case CBOR.deserialiseFromBytes decodeDerPath (BL.fromStrict bytes) of
+        Right (_, result) -> pure result
+        _ -> fail "decodeDerivationPath: unable to decode path"
+
+
+decodeDerPath :: CBOR.Decoder s (Index 'Hardened 'AccountK, Index 'Soft 'AddressK)
+decodeDerPath = decodeListIndef CBOR.decodeWord32 >>= \case
+    [accIx, addrIx] -> pure (Index accIx, Index addrIx)
+    ps -> fail $ "decodeDerPath: Unexpected derivation path length ("
+        ++ (show $ length ps) ++ ")"
+
+  where
+    -- | Decode an arbitrary long list. CBOR introduce a "break" character to
+    -- mark the end of the list, so we simply decode each item until we encounter
+    -- a break character.
+    --
+    -- Example:
+    --
+    --     myDecoder :: CBOR.Decoder s [MyType]
+    --     myDecoder = decodeListIndef decodeOne
+    --       where
+    --         decodeOne :: CBOR.Decoder s MyType
+    --
+    decodeListIndef :: forall s a. CBOR.Decoder s a -> CBOR.Decoder s [a]
+    decodeListIndef decodeOne = do
+        _ <- CBOR.decodeListLenIndef
+        CBOR.decodeSequenceLenIndef (flip (:)) [] reverse decodeOne
+
+-- | Derive a symmetric key for encrypting and authenticating the address
+-- derivation path.
+hdPassphrase :: Key 'RootK XPub -> Passphrase "address-der"
+hdPassphrase (Key rootXPub) = Passphrase $
+    PBKDF2.generate
+    (PBKDF2.prfHMAC SHA512)
+    -- fixme: check if enough rounds -- if running on a public key, it probably doesn't matter.
+    (PBKDF2.Parameters 500 32)
+    (unXPub rootXPub)
+    ("address-hashing" :: ByteString)
+
+{-------------------------------------------------------------------------------
+                    HD payload encryption and authentication
+-------------------------------------------------------------------------------}
+
+cardanoNonce :: ByteString
+cardanoNonce = "serokellfore"
+
+-- | ChaCha20/Poly1305 encrypting and authenticating the HD payload of addresses.
+encryptDerPath
+    :: Passphrase "address-der"
+       -- ^ Symmetric key / passphrase, 32-byte long
+    -> ByteString -- Payload to be encrypted
+    -> CryptoFailable ByteString -- Ciphertext with a 128-bit crypto-tag appended.
+encryptDerPath (Passphrase passphrase) payload = do
+    nonce <- Poly.nonce12 cardanoNonce
+    st1 <- Poly.finalizeAAD <$> Poly.initialize passphrase nonce
+    let (out, st2) = Poly.encrypt payload st1
+    return $ out <> BA.convert (Poly.finalize st2)
+
+decryptDerPath
+    :: Passphrase "address-der"
+       -- ^ Symmetric key / passphrase, 32-byte long
+    -> ByteString -- Payload to be encrypted
```

Comment:
don't we want to use here Address to better seld-document decryptDerPath function?

Reply:
I considered it, but this string is just a part of the address.